Add blog article about gRPC probes beta (#32664)
kubernetes/website: content/en/blog/_posts/2022-04-15-grpc-probes-in-beta.md

---
layout: blog

> title: "Try the new gRPC probes in 1.24: now in beta."

Reviewer: ```suggestion
title: "Try the new gRPC probes in 1.24: now in beta"
```

date: 2022-04-15
slug: grpc-probes-now-in-beta
---

> 

Reviewer: Mention an author (see other articles for an example).

With Kubernetes 1.24 the gRPC probes functionality entered beta and is available by default.
Now you can configure startup, liveness, and readiness probes for your gRPC app

> without exposing http endpoint or using an extra executable. All built-in to the Kubernetes.

Reviewer: ```suggestion
without exposing any HTTP endpoint, nor do you need an executable. Kubernetes can natively
connect to your your workload via gRPC and query its status.
```

## Some history

> Kubernetes allows to control the container lifecycle from inside the app
> by checking http response, running executable, or simply connecting to a tcp port.

Reviewer: ```suggestion
It's useful to let the system managing your workload check that the app is healthy,
has started OK, and whether the app considers itself good to accept traffic.
Before the gRPC support was added, Kubernetes already allowed you to check
for health based on running an executable from inside the container image,
by making an HTTP request, or by checking whether a TCP connection succeeded.
```

> Those checks are often enough for the most of apps. For gRPC apps, it is easy

Reviewer: ```suggestion
For most apps, those checks are enough. If your provides a gRPC endpoint for a health
(or readiness) check, it is easy
```

> to repurpose the `exec` probe to use it for gRPC health checking. This solution
> was described in the blog post "[Health checking gRPC servers on Kubernetes](https://kubernetes.io/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/)"
> and used for a long time. Commonly used tool to enable this was created on [Aug 21, 2018](https://github.com/grpc-ecosystem/grpc-health-probe/commit/2df4478982e95c9a57d5fe3f555667f4365c025d) with

Reviewer: ```suggestion
to repurpose the `exec` probe to use it for gRPC health checking.
In the blog article [Health checking gRPC servers on Kubernetes](/blog/2018/10/01/health-checking-grpc-servers-on-kubernetes/), Ahmet Alp Balkan described how you can do that—a
mechanism that still works today.

There is a commonly used tool to enable this that was [created](https://github.com/grpc-ecosystem/grpc-health-probe/commit/2df4478982e95c9a57d5fe3f555667f4365c025d)
on August 21, 2018, and with
```

the first release at [Sep 19, 2018](https://github.com/grpc-ecosystem/grpc-health-probe/releases/tag/v0.1.0-alpha.1).

This approach for gRPC apps health checking is very popular. There are [3,626 Dockerfiles](https://github.com/search?l=Dockerfile&q=grpc_health_probe&type=code)
with the `grpc_health_probe` and [6,621 yaml](https://github.com/search?l=YAML&q=grpc_health_probe&type=Code) files that are discovered with the
basic search on GitHub (at the moment of writing). This is good indication of the tool popularity
and the need to support this natively.

> With the Kubernetes 1.23, native support for gRPC health probing was introduced as alpha feature.
> The feature is not at beta and enabled by default.

Reviewer: ```suggestion
Kubernetes v1.23 introduced an alpha-quality implementation of native support for
querying a workload status using gRPC.
Because it was an alpha feature, this was disabled by default for the v1.23 release.
```

## Using the feature

> The gRPC health checking is [easy to use](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe).

Reviewer: You can reword this in two ways:
- express _your opinion_ that it is easy to use (opinions are OK in blog articles)
- use neutral language

not checking the server certificates and ignore related problems.

Built-in checks also cannot be configured to ignore certain types of errors
(`grpc_health_probe` returns different exit codes for different errors),
and cannot be "chained" to run the health check on multiple services in a single probe.

But all these limitations are quite standard for gRPC and there are easy workarounds
for those.

## Try it for yourself

### Getting cluster

You can try this feature today. To try the feature you can spin up the kubernetes cluster
yourself with the `GRPCContainerProbe` feature gate enabled.

Alternatively, at the moment of writing, you can spin up the cluster on GKE
using the following command (note, version is `1.23` and `enable-kubernetes-alpha` are specified):

``` bsh
$ gcloud container clusters create test-grpc \
    --enable-kubernetes-alpha \
    --no-enable-autorepair \
    --no-enable-autoupgrade \
    --release-channel=rapid \
    --cluster-version=1.23
```

You will also need to configure `kubectl` to access the cluster:

``` bsh
$ gcloud container clusters get-credentials test-grpc

> ```

Reviewer: - we should explain how to get a v1.24 cluster, not a v1.23 cluster
- we shouldn't require or imply that the user uses Google Cloud

### Trying the feature out

> Lets create the pod to test how gRPC probes work. The `agnhost` image has

Reviewer: What's an `agnhost`? I imagine readers will ask that exact question.

a useful [grpc-health-checking](https://github.com/kubernetes/kubernetes/blob/b2c5bd2a278288b5ef19e25bf7413ecb872577a4/test/images/agnhost/README.md#grpc-health-checking) module
that exposes two ports - one is serving health checking service,
another - http port to react on commands `make-serving` and `make-not-serving`.

Here is an example pod definition. It starts the `grpc-health-checking` module,
exposes ports `5000` and `8080`, and configures gRPC readiness probe:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-grpc
spec:
  containers:
  - name: agnhost
    image: k8s.gcr.io/e2e-test-images/agnhost:2.35
    command: ["/agnhost", "grpc-health-checking"]
    ports:
    - containerPort: 5000
    - containerPort: 8080
    readinessProbe:
      grpc:
        port: 5000
```

If the file called `test.yaml`, you can create the pod and check it's status.
The pod will be in ready state as indicated by the snippet of the output.

``` bsh
$ kubectl apply -f test.yaml
$ kubectl describe test-grpc

...
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True

> ...

Reviewer: This isn't a valid shell script. Consider following how the docs do it: [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

```

Now let's change the health checking endpoint status to NOT_SERVING.
In order to call the http port of the Pod, let's create a port forward:

``` bsh

> $ kubectl port-forward test-grpc 8080:8080

Reviewer: Don't include a prompt; it makes the code not be valid shell.

```

You can `curl` to call the command...

``` bsh

> $ curl http://localhost:8080/make-not-serving

Reviewer: Don't include a prompt; it makes the code not be valid shell.
Also, the highlight for POSIX shell is `shell` not `bsh`.

```

... and in a few seconds the port status will switch to not ready.

``` bsh
$ kubectl describe test-grpc

Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True

...

  Warning  Unhealthy  2s (x6 over 42s)  kubelet            Readiness probe failed: service unhealthy (responded with "NOT_SERVING")
```

> Once it is switched back, in ~1s the port will get back to ready status:

Reviewer: ```suggestion
Once it is switched back, in about one second the port will get back to ready status:
```
Did you mean “Pod” instead of “port”?

``` bsh
$ curl http://localhost:8080/make-serving
$ kubectl describe test-grpc

...

Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
```

> Using gRPC health probing on Kubernetes is easy and straightforward. Follow the [documentation to learn more](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe)

Reviewer: ```suggestion
This new built-in gRPC health probing on Kubernetes makes implementing ahealth-check via gRPC
much easier than the older approach that relied on using a separate `exec` probe. Read through
the official
[documentation](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe)
to learn more.
```

Exec probes are generally slower than a gRPC call as they require instantiating a new process to run an executable.
It also makes the checks less sensible for edge cases when the pod is running at maximum resources and has troubles
instantiating new processes.

There are a few limitations though. Since configuring a client certificate for probes is hard,
services that require client authentication are not supported. The built-in probes are also
not checking the server certificates and ignore related problems.

Built-in checks also cannot be configured to ignore certain types of errors
(`grpc_health_probe` returns different exit codes for different errors),
and cannot be "chained" to run the health check on multiple services in a single probe.

But all these limitations are quite standard for gRPC and there are easy workarounds
for those.

## Try it for yourself

### Getting cluster

You can try this feature today. To try the feature you can spin up the kubernetes cluster
yourself with the `GRPCContainerProbe` feature gate enabled, there are many [tools available](/docs/tasks/tools/).

> Alternatively, at the moment of writing, you can spin up the cluster on GKE for a quick test.

Reviewer 2: Can we say that one needs to bring up a cluster on 1.24 and then for e.g. this is how you would do it on GKE. Also, mention that it should work for kind or any other 1.24 based kubernetes clusters.

Author: Added a paragraph

Other vendors may also have similar capabilities, especially if you
read this blog post long after kubernetes 1.24 released.

On GKE use the following command (note, version is `1.23` and `enable-kubernetes-alpha` are specified).

``` bsh
gcloud container clusters create test-grpc \
    --enable-kubernetes-alpha \
    --no-enable-autorepair \
    --no-enable-autoupgrade \
    --release-channel=rapid \
    --cluster-version=1.23
```

You will also need to configure `kubectl` to access the cluster:

``` bsh
gcloud container clusters get-credentials test-grpc
```

### Trying the feature out

> Lets create the pod to test how gRPC probes work. For this test we will use the `agnhost` image.

Reviewer 2: nit: Let's

Author: applied

This is a k8s maintained image with that can be used for all sorts of workload testing.
For example, it has a useful [grpc-health-checking](https://github.com/kubernetes/kubernetes/blob/b2c5bd2a278288b5ef19e25bf7413ecb872577a4/test/images/agnhost/README.md#grpc-health-checking) module
that exposes two ports - one is serving health checking service,
another - http port to react on commands `make-serving` and `make-not-serving`.

Here is an example pod definition. It starts the `grpc-health-checking` module,
exposes ports `5000` and `8080`, and configures gRPC readiness probe:

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-grpc
spec:
  containers:
  - name: agnhost
    image: k8s.gcr.io/e2e-test-images/agnhost:2.35
    command: ["/agnhost", "grpc-health-checking"]
    ports:
    - containerPort: 5000
    - containerPort: 8080
    readinessProbe:
      grpc:
        port: 5000
```

If the file called `test.yaml`, you can create the pod and check it's status.
The pod will be in ready state as indicated by the snippet of the output.

> ``` bsh

Reviewer 2: Is this supposed to be `bash`?

Author: changed to ```shell for consistency with other docs

```

Now let's change the health checking endpoint status to NOT_SERVING.
In order to call the http port of the Pod, let's create a port forward:

``` bsh
kubectl port-forward test-grpc 8080:8080
```

You can `curl` to call the command...

``` bsh
curl http://localhost:8080/make-not-serving
```

... and in a few seconds the port status will switch to not ready.

``` bsh
kubectl describe test-grpc
```

The output now will have:

```
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True

...

  Warning  Unhealthy  2s (x6 over 42s)  kubelet            Readiness probe failed: service unhealthy (responded with "NOT_SERVING")
```

Once it is switched back, in about one second the Pod will get back to ready status:

``` bsh
curl http://localhost:8080/make-serving
kubectl describe test-grpc
```

And output indicates that service went back to ready:

```
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
```

This new built-in gRPC health probing on Kubernetes makes implementing a health-check via gRPC

> much easier than the older approach that relied on using a separate `exec` probe. Read through

Reviewer 2: what about `easier and reduces overhead compared..`

---
layout: blog

> title: "Try the new gRPC probes in 1.24: now in beta"

Reviewer: ```suggestion
title: "Kubernetes 1.24: gRPC container probes in beta"
```
(and adjust the slug to match)

> date: 2022-04-15

Reviewer 3: Publish date should be after 2022-03-05 since this blog is a feature blog for the 1.24 release

Author: ```suggestion
date: 2022-05-05
```

> date: 2022-05-05

Reviewer: ```suggestion
date: 2022-05-13
```
Please update the path to match; it doesn't affect the published article but it does help contributors.

if you are familiar with other probe types in Kubernetes.
The natively supported health probe has many benefits over the workaround involving `grpc_health_probe` executable.

With the native gRPC support you don't need to download and carry `10MB` of an additional executable with your image.
Exec probes are generally slower than a gRPC call as they require instantiating a new process to run an executable.
It also makes the checks less sensible for edge cases when the pod is running at maximum resources and has troubles
instantiating new processes.

There are a few limitations though. Since configuring a client certificate for probes is hard,
services that require client authentication are not supported. The built-in probes are also
not checking the server certificates and ignore related problems.

Built-in checks also cannot be configured to ignore certain types of errors
(`grpc_health_probe` returns different exit codes for different errors),
and cannot be "chained" to run the health check on multiple services in a single probe.

But all these limitations are quite standard for gRPC and there are easy workarounds
for those.

## Try it for yourself

> ### Getting cluster

Reviewer: ```suggestion
### Cluster-level setup
```

> You can try this feature today. To try the feature you can spin up the kubernetes cluster

Reviewer: ```suggestion
You can try this feature today. To try native gRPC probes, you can spin up a Kubernetes cluster
```

yourself with the `GRPCContainerProbe` feature gate enabled, there are many [tools available](/docs/tasks/tools/).

Since the feature gate `GRPCContainerProbe` is enabled by default in 1.24,
many vendors will have this functionality working out of the box.
So you may just create an 1.24 cluster on platform of your choice. Some vendors
allow to enable alpha features on 1.23 clusters.

For example, at the moment of writing, you can spin up the test cluster on GKE for a quick test.
Other vendors may also have similar capabilities, especially if you

> read this blog post long after kubernetes 1.24 released.

Reviewer: ```suggestion
are reading this blog post long after the Kubernetes 1.24 release.
```

On GKE use the following command (note, version is `1.23` and `enable-kubernetes-alpha` are specified).

```shell
gcloud container clusters create test-grpc \
    --enable-kubernetes-alpha \
    --no-enable-autorepair \
    --no-enable-autoupgrade \
    --release-channel=rapid \
    --cluster-version=1.23
```

You will also need to configure `kubectl` to access the cluster:

```shell
gcloud container clusters get-credentials test-grpc

> ```

Reviewer: I'd definitely prefer an example that wasn't cloud-provider-specific. Nudging readers towards GKE isn't the right look.

It's OK to provide a _set_ of examples; for example, you could have examples for:
- GKE
- kind
- minikube

(we prefer to list in alphabetical order)

Author: nudging wasn't an intention, I just wanted to provide something easier than kind to make people try it. Kind instructions will be a bit more involved that I'd prefer - changing feature gates will be needed.

if this is not ideal, I can remove that part completely only leaving the references to `kind`. Not sure if providing step-by-step instruction would be useful for anybody. People who will try it with kind, likely already know it. People who doesn't - don't need those.

wdyt?

Reviewer: I'm fine with either option.

### Trying the feature out

Let's create the pod to test how gRPC probes work. For this test we will use the `agnhost` image.
This is a k8s maintained image with that can be used for all sorts of workload testing.
For example, it has a useful [grpc-health-checking](https://github.com/kubernetes/kubernetes/blob/b2c5bd2a278288b5ef19e25bf7413ecb872577a4/test/images/agnhost/README.md#grpc-health-checking) module
that exposes two ports - one is serving health checking service,
another - http port to react on commands `make-serving` and `make-not-serving`.

Here is an example pod definition. It starts the `grpc-health-checking` module,
exposes ports `5000` and `8080`, and configures gRPC readiness probe:

``` yaml

> apiVersion: v1

Reviewer: ```suggestion
---
```

I tend to start YAML examples with `---`. It's not required but visually it helps people spot that they're reading YAML.

Author: did you mean to remove the apiVersion in code proposal?

Author: I cannot find many examples at kubernetes.io where the snippets starts with `---`, but happy to apply if it's best

Reviewer: We used to have a bug where we couldn't process manifests that started `---` but (fortunately) that's gone now.

kind: Pod
metadata:
  name: test-grpc
spec:
  containers:
  - name: agnhost
    image: k8s.gcr.io/e2e-test-images/agnhost:2.35
    command: ["/agnhost", "grpc-health-checking"]
    ports:
    - containerPort: 5000
    - containerPort: 8080
    readinessProbe:
      grpc:
        port: 5000
```

If the file called `test.yaml`, you can create the pod and check it's status.
The pod will be in ready state as indicated by the snippet of the output.

```shell
kubectl apply -f test.yaml
kubectl describe test-grpc
```

The output will contain something like this:

```
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
```

Now let's change the health checking endpoint status to NOT_SERVING.
In order to call the http port of the Pod, let's create a port forward:

```shell
kubectl port-forward test-grpc 8080:8080
```

You can `curl` to call the command...

```shell
curl http://localhost:8080/make-not-serving
```

... and in a few seconds the port status will switch to not ready.

```shell

> kubectl describe test-grpc

Reviewer: Did you mean:
```suggestion
kubectl describe pod test-grpc
```

```

The output now will have:

```
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True

...

  Warning  Unhealthy  2s (x6 over 42s)  kubelet            Readiness probe failed: service unhealthy (responded with "NOT_SERVING")
```

Once it is switched back, in about one second the Pod will get back to ready status:

``` bsh
curl http://localhost:8080/make-serving
kubectl describe test-grpc
```

> And output indicates that service went back to ready:

Reviewer: ```suggestion
The output indicates that the Pod went back to being `Ready`:
```

```
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
```

This new built-in gRPC health probing on Kubernetes makes implementing a health-check via gRPC
much easier than the older approach that relied on using a separate `exec` probe. Read through
the official
[documentation](/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe)
to learn more and provide feedback before the feature will be promoted to GA.

> # Summary

Reviewer: ```suggestion
## Summary
```
We don't use first-level headings within Markdown.
